### PR TITLE
GH-296: Extended the protocol with `textDocument/callHierarchy`.

### LINKED ISSUE #296: Support for call hierarchies
The proposed LSP extension is [here](https://github.com/Microsoft/vscode-languageserver-node/pull/420).

### PR DESCRIPTION
**Pull request by kittaakos**
Closes #296.

Signed-off-by: Akos Kitta <kittaakos@typefox.io>

----

LSP PR: https://github.com/Microsoft/vscode-languageserver-node/pull/420

### CODE REVIEW

**Comment by kittaakos** [comment]
> should we also add some remark in their JavaDoc comments?

I will add some additional notes to the Javadoc.

(Will do the same for the [`textDocument/typeHierarchy` related changes](https://github.com/eclipse/lsp4j/pull/273) too.)

**Comment by kittaakos** [comment]
I have documented the caveats between LSP4J and LSP for the following language features:
 - `textDocument/semanticHighlighting`,
 - `textDocument/typeHierarchy`, and
 - `textDocument/callHierarchy`.

**Comment by spoenemann** [comment]
Great! :+1:

**Comment by kittaakos** [comment]
I had to update the `typeHierarchy` API based on the LSP [feedback](https://github.com/Microsoft/vscode-languageserver-node/pull/426#pullrequestreview-196406599).

File: org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/TypeHierarchyDirection.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+ * <p>
+ * Valid values are:
+ * <ul>
+ * <li>{@link TypeHierarchyDirection#Children Children},</li>
+ * <li>{@link TypeHierarchyDirection#Parents Parents},</li>
+ * <li>{@link TypeHierarchyDirection#Both Both}.</li>
+ * </ul>
+ *
+ */
+public enum TypeHierarchyDirection {
```
Comment (spoenemann) [inline org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/TypeHierarchyDirection.java]: The enum needs a field 'value' for the conversion from/to numbers to work, see other enums in the project.

The conversion is defined in [EnumTypeAdapter](https://github.com/eclipse/lsp4j/blob/master/org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/adapters/EnumTypeAdapter.java).

Comment (kittaakos) [reply org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/TypeHierarchyDirection.java]: Hold up a sec, what about [`TextDocumentSyncKind`](https://github.com/eclipse/lsp4j/blob/db9f9bdda89a67384f5c78985b4d3f89a001a92d/org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/TextDocumentSyncKind.java#L17). My understanding was, we need the `value` field if the enum starts from a **non-zero** number bu we can use zero if the enum is optional on the use site. Can you please clarify? Thanks!

Comment (spoenemann) [reply org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/TypeHierarchyDirection.java]: Right, the type adapter falls back to the ordinal value. Never mind, then.

**Comment by spoenemann** [comment]
:+1:

**Pull request merged**